### PR TITLE
Feat/issue 61

### PR DESCRIPTION
**Pull request by math4humanities**
Fixes #
Design discussion issue #

## Changes

Please provide a brief description of the changes here.
Added story volume alert function using celery beat. Transformed processor tasks into a package, where tasks and alerts modules were separated successfully. Unit test for story volume alert was added. 
## Merge requirement checklist

* [ ] Unit tests added/updated
* [ ] Doc-strings & documentation added/updated

### CODE REVIEW

**Comment by math4humanities** [comment]
Figure out why pytest works locally but not here

**Comment by math4humanities** [comment]
Changed the module name, and reverted the requirements!

File: processor/tasks/alerts.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    # Get total stories processed over the past 4 days
+    earliest_date = dt.datetime.now() - dt.timedelta(days=days)
+    result = session.execute(
+        select(func.count(Story.id)).where(Story.processed_date >= earliest_date)
+    )
+    total_stories = result.scalar()
+    return total_stories
+
+
+def calculate_average_stories(total_stories: int, days: int) -> float:
```
Comment (rahulbot) [inline processor/tasks/alerts.py:44]: Making this a function feels like a little bit of over-engineering.  I get that you want to test the logic in a unit test, but it isn't the kind of simple logic that I think requires that level of testing. I'd remove the function and the unit test and just do the division inline.

File: processor/tasks/alerts.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+
+def send_alert(total_stories: int, days: int, threshold: float):
+    # Get average stories over n days, in our case n=4
+    average_stories_per_day = calculate_average_stories(total_stories, days)
+    logger.info(
+        f"Average stories processed per day in the last {days} days: {average_stories_per_day:.1f}"
+    )
+
+    warning_message = (
```
Comment (rahulbot) [inline processor/tasks/alerts.py:55]: I'd move this inside the `if` below, because you only use the string if the condition is met.

File: processor/tasks/alerts.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        f"Warning: Average stories per day in the last {days} days is {average_stories_per_day:.1f}, "
+        f"which is far below the expected value of >= {threshold}."
+    )
+    # Check if it's below threshold and send messages/emails
+    if average_stories_per_day < threshold:
+        logger.warning(warning_message)
+
+        # Check if Slack is configured before sending message
+        if is_slack_configured():
```
Comment (rahulbot) [inline processor/tasks/alerts.py:64]: Call to `is_slack_configured` is redundant - `send_slack_msg` makes the call inside already to be safe.
https://github.com/dataculturegroup/feminicide-story-processor/blob/2fdfdc5cc8146b37bc56449a5ffb64a8a8fca6ec/processor/notifications.py#L74
I'd suggest removing this `if` clause.

Note: `send_slack_msg` seems to return `False` if it is _not_ configured, but not `True` if it is :-( Can you make that small change in here as well to have it `return True` at the end of the function, and add the typing hint accordingly?

File: processor/tasks/alerts.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+                slack_config["bot_token"],
+                "Story Fetcher",
+                "Low Story Count Alert",
+                warning_message,
+            )
+        else:
+            logger.warning("Slack is not configured.")
+
+        # Check if email is configured before sending
+        if is_email_configured():
```
Comment (rahulbot) [inline processor/tasks/alerts.py:77]: Same note here - `send_email` does the `is_email_configured` check for you. Note how it _does_ return `True` at the end if it was sent (`False` if unconfigured).

File: processor/tasks/alerts.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+            logger.warning("Email is not configured.")
+
+
+@app.task
+def check_story_count():
+    """
+    Task to check the average number of stories processed over the last 4 days and send alerts if below threshold.
+    """
+
+    days = 4
```
Comment (rahulbot) [inline processor/tasks/alerts.py:92]: I'd make this a constant at the top of the file: `STORY_COUNT_WINDOW_SIZE = 4` or something.

File: processor/tasks/alerts.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    threshold = 40000
```
Comment (rahulbot) [inline processor/tasks/alerts.py:93]: I'd make this a constant at the top of the file: `STORY_COUNT_THRESHOLD = 40000` or something, and add a comment indicating that it was based on looking at data in July 2024.

File: processor/tasks/alerts.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+)
+fileHandler = logging.FileHandler(
+    os.path.join(path_to_log_dir, "tasks-{}.log".format(time.strftime("%Y%m%d-%H%M%S")))
+)
+fileHandler.setFormatter(logFormatter)
+logger.addHandler(fileHandler)
+
+
+def get_total_stories_over_n_days(
+    session: database.get_session_maker(), days: int = 4
```
Comment (rahulbot) [inline processor/tasks/alerts.py:33]: I worry about having a default session allowed here. DB sessions can cause issues when they aren't intentionally opened and closed, especially in parallelized apps like this. I'd remove that default value for `session`.

File: processor/tasks/test/test_alerts.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+class TestAlerts(unittest.TestCase):
+
+    def test_calculate_average_stories(self):
+        result = alerts.calculate_average_stories(20, 5)
+        self.assertEqual(result, 4.0)
+
+    @patch("processor.database.get_session_maker")
+    def test_get_total_stories_over_n_days(self, mock_get_session_maker):
+        mock_session = MagicMock()
```
Comment (rahulbot) [inline processor/tasks/test/test_alerts.py:15]: Honestly never used `MagicMock` and `patch` before - thanks for introducing me to such a useful a new pattern!

File: processor/tasks/test/test_alerts.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        self.assertEqual(total_stories, 20000)
+
+    @patch("processor.tasks.alerts.send_email")
+    @patch("processor.tasks.alerts.send_slack_msg")
+    @patch("processor.tasks.alerts.logger")
+    @patch("processor.tasks.alerts.is_email_configured")
+    @patch("processor.tasks.alerts.is_slack_configured")
+    @patch("processor.tasks.alerts.get_slack_config")
+    @patch("processor.tasks.alerts.get_email_config")
+    def test_send_alert(
```
Comment (rahulbot) [inline processor/tasks/test/test_alerts.py:32]: I think this is testing the positive case (value is below threshold). Can you add a test for the negative value as well (value is above threshold, so no alerts sent)?

File: processor/tasks/alerts.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+    with session() as session:
+        total_stories = get_total_stories_over_n_days(session, days)
+        send_alert(total_stories, days, threshold)
+
+
+# Configuration to schedule the check_story_count task
+app.conf.beat_schedule = {
+    "check_story_count": {
+        "task": "processor.tasks.alerts.check_story_count",
+        "schedule": crontab(hour="2", minute="0", day_of_month="*/4"),
```
Comment (rahulbot) [inline processor/tasks/alerts.py:106]: I think we want this to run every day, so it is using a rolling avg of prior 4 days.

**Review by rahulbot — CHANGES_REQUESTED** [review]
Looking good 👍🏽 Left various minor notes.
Also note that `.pre-commit-config.yaml` and `requirements.txt` still show some changes, but I'd like them to have no diffs on this branch so merging this PR and #64 are well separated.

**Review by rahulbot — APPROVED** [review]
Looks good. I think this is ready to merge.
